clang/lib/CIR/Dialect/Transforms/LifetimeCheck.cpp

```diff
@@ -1604,10 +1927,29 @@ void LifetimeCheckPass::checkCtor(CallOp callOp, cir::CXXCtorAttr ctor) {
 void LifetimeCheckPass::checkOperators(CallOp callOp,
                                        ASTCXXMethodDeclInterface m) {
   auto addr = getThisParamOwnerCategory(callOp);
+
   if (addr) {
-    // const access to the owner is fine.
+    // Smart pointers need special handling even for const methods
+    // (operator*, operator-> are const but unsafe when null)
+    if (isSmartPointerType(addr.getType())) {
+      // Get declaration name (works for operators too)
+      std::string methodName = m.getDeclName().getAsString();
+
+      // Safe methods: get, release, reset, operator bool
+      if (methodName == "get" || methodName == "release" ||
```

Comment:
This check should be in its own helper function, `isSmartPointerSafeMethod` or something akin